keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -263,35 +280,57 @@ is scoped to.
-scalar data element the validator is scoped to. The data element will be accessible to CEL
-expressions via `self`, e.g. `len(self) > 10`.
+  scalar data element the validator is scoped to. The data element will be accessible to CEL
+  expressions via `self`, e.g. `len(self) > 10`.
 
 - For OpenAPIv3 list and map types, the expression will have access to the data element of the list
 or map. These will be accessible to CEL via `self`. The elements of a map or list can be validated using the CEL support for collections
 like the `all` macro, e.g. `self.all(listItem, <predicate>)` or `self.all(mapKey,
 <predicate>)`.
   
 - For immutability use case, validator will have access to the existing version of the object. This
-  will be accessible to CEL via the `old<propertyName>` identifier.
+  will be accessible to CEL via the `oldSelf` identifier.
   - This will only be available on mergable collection types such as objects (unless
     `x-kubernetes-map-type=atomic`), maps with `x-kubernetes-map-type=granular` and lists
     with `x-kubernetes-list-type` set to `set` or `map`.  See [Merge
     Strategy](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy) for
     details.
   - The use of "old" is congruent with how `AdmissionReview` identifies the existing object as
-    `oldObject`. To avoid name collisions `old<propertyName>` will be treated the same as a CEL
-    keyword for escaping purposes (see below).
+    `oldObject`.
   - xref [analysis of possible interactions with immutability and
     validation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1101-immutable-fields#openapi-extension-x-kubernetes-immutable).
 
-- If a object property collides with a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
-  it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision, all properties with a `_` prefix will always be
+- If a object property name is a CEL keyword (see RESERVED in [CEL Syntax](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax)),
+  it will be escaped by prepending a _ prefix. To prevent this from causing a subsequent collision,  properties named with a CEL keyword and a  `_` prefix will be
   prefixed by `__` (generally, N+1 the existing number of `_`s).
 
+- Only property names of the form `[a-zA-Z_.-/][a-zA-Z0-9_.-/]*` are accessible and are escaped
+  according to the following rules when accessed in the expression:
+	- '__' escapes to '__underscores__'
+	- '.' escapes to '__dot__'
+	- '-' escapes to '__dash__'
+	- '/' escapes to '__slash__'
+	- Property names that match a CEL RESERVED keyword exactly escape to '__{keyword}__'. The
+	keywords are: "true", "false", "null", "in", "as", "break", "const", "continue", "else", "for",
+	"function", "if", "import", "let", "loop", "package", "namespace", "return".
```

Comment:
I'm confused by this line. A few lines up says:

> If a object property name is a CEL keyword (see RESERVED... ) it will be escaped by prepending a _ prefix."

These both seem to refer to the same RESERVED keywords and/but layout different spellings for them. Do I understand correctly?

Does the rule on an object with the `package` property refer to it using `self._package` or `self.__package__`?

Reply:
`__package__` is correct. The above example is out-of-date from a previous design iteration. I'll remove it now.